### PR TITLE
[AArch64] Port atomic rmw to ISLE

### PR DESCRIPTION
**Pull request by sparker-arm**
Also fix the current implementation, as AArch64
AtomicRMWOp::Clr != AtomicRmwOp::And, as the input needs to be
inverted first. We now also lower Xchg to Swp and Sub to Add
with a negated input.

Copyright (c) 2022, Arm Limited.

### CODE REVIEW

**Comment by sparker-arm** [comment]
Ah, the joys of non-native development... Looks like I need to fix the loads for the signed operations.

**Comment by cfallin** [comment]
I just rebased this PR over the recent RA2 merge to fix up the conflicts. Sorry about the noise and happy to review this once the signed bits are fixed!

**Comment by sparker-arm** [comment]
Sorry, I made a mess of my rebase and so now I've squashed everything back into one commit.

File: cranelift/codegen/src/isa/aarch64/inst.isle [hunk 1]
```diff
@@ -2053,3 +2096,30 @@
...
+
+;; Helper for emitting `MInst.AtomicRMW` instructions.
+(decl lse_atomic_rmw (AtomicRMWOp Value Reg Type) Reg)
+(rule (lse_atomic_rmw op p r_arg2 ty)
+      (let (
+          (r_addr Reg p)
+          (dst WritableReg (temp_writable_reg ty))
+          (_ Unit (emit (MInst.AtomicRMW op r_arg2 dst r_addr ty)))
+        )
+        (writable_reg_to_reg dst)))
```
Comment (cfallin) [inline cranelift/codegen/src/isa/aarch64/inst.isle:2108]: The `writable_reg_to_reg` is unnecessary now with implicit conversions (harmless, just more verbose than needed).

**Review by cfallin — APPROVED** [review]
Thanks! One miniscule nit below, happy to merge with that updated.

File: cranelift/codegen/src/isa/aarch64/inst.isle (hunk 1 above)
Comment (akirilov-arm) [reply cranelift/codegen/src/isa/aarch64/inst.isle]: @cfallin I depend on this PR in #4034, so both @sparker-arm and I would appreciate any feedback on the ISLE code - it should look good now.

**Review by cfallin — APPROVED** [review]

**Comment by cfallin** [comment]
Looks good, thanks!

**Pull request merged**